### PR TITLE
Implement PReLU in a compositional way

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91238

The PReLU implementation was all over the place. This lead to a number
of bugs like https://github.com/pytorch/pytorch/issues/68760.  We fix it by:
- Keeping the weird broadcasting logic it has as a CompositeImplicit kernel that calls into a second kernel
- This second kernel is just a good-ol' pointwise kernel.
- We implement the derivative for the pointwise kernel via TI as well for speed.
- We implement the second derivative for the pointwise kernel and the forward AD derivatives compositionally

This fixes a number of issues:
- We don't perform copies any more when the inputs are not contiguous
- The derivatives are now correct
- We fix vmap and many other functorch-related issues.
- CPU and CUDA now share the relevant broadcasting logic
- The implementation is about 1/3 the length.

Fixes https://github.com/pytorch/pytorch/issues/68760
Fixes https://github.com/pytorch/pytorch/issues/89895

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by lezcano** [comment]
@soulitzer @albanD you have been tagged as codeowners, but could you please review it?

The only issue with the current approach is that I was not able to port the MKLDNN and the MPS implementations to just implement the simpler new kernel, so those should still be buggy. If the relevant PoCs from Intel / Apple port these, we should be able to remove all the relevant entries from derivatives.yaml and native_functions.yaml

File: aten/src/ATen/native/native_functions.yaml
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:4658]: cc @kulinseth if you are interested in this one.

**Review by albanD — COMMENTED** [review]
I'm not sure why this has to be more complex than functions like amax which we can implement without a pair a functions?

**Review by kshitij12345 — APPROVED** [review]
LGTM (on functorch batching rule front) :)

File: aten/src/ATen/native/Activation.cpp
Comment (ngimel) [inline aten/src/ATen/native/Activation.cpp:710]: I'm not sure we want type promoting behavior for `prelu`, and we don't have it currently.

File: torch/_decomp/decompositions.py
Comment (ngimel) [inline torch/_decomp/decompositions.py:290]: pw_cast_for_opmath not needed here, it only leads to more complicated IR

File: torch/_decomp/decompositions.py
Comment (ngimel) [inline torch/_decomp/decompositions.py:296]: similarly here

**Comment by albanD** [comment]
I'm not sure why the fact that you have a custom backward changes this? Is it just because you want a simpler backward function?
But the backward kernel already doesn't need to handle broadcasting (which is handled in the engine).

If this implementation works, no need to re-do it again. Just wanted to know why we're going this way.

File: torch/_decomp/decompositions.py
Comment (ngimel) [reply torch/_decomp/decompositions.py:290]: We are outputting either `self` or `weight * self`, in both cases there's no difference whether they are pre-converted to float or not here - `self` is a copy either way, `weight * self` would do whatever is needed in multiplication

Comment (lezcano) [reply torch/_decomp/decompositions.py:290]: I'll clean up all that code then.

**Comment by jbschlosser** [comment]
> Quantized fails as well, no idea why :(

Registering `QuantizedCPU: prelu_quantized_cpu` underneath `_prelu_kernel` instead of `prelu` and removing the now-redundant weight-shaping logic fixes the tests for me:
https://github.com/pytorch/pytorch/blob/67c53d50e5b34a536cce2ad98b9ca5f65ff8a34d/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L683-L690

But I'm not exactly clear on why the dispatcher is decomposing `prelu` -> `_prelu_kernel` instead of calling the registered `QuantizedCPU` in your current PR.

**Comment by lezcano** [comment]
Thank you for the pointer @jbschlosser! I fixed that but now I'm getting an error, but I don't understand the quantised code so I don't really know what's going on there. Perhaps @Xia-Weiwen can have a look?

The MKLDNN part was written by @yanbing-j. Could you please tell me how could we fix that bit?

**Comment by jbschlosser** [comment]
> I fixed that but now I'm getting an error, but I don't understand the quantised code so I don't really know what's going on there.

@lezcano what's the error you're getting?

**Comment by lezcano** [comment]
The issue in quantized is the one that's in https://github.com/pytorch/pytorch/actions/runs/3795264825/jobs/6454831858 @Xia-Weiwen any idea what could be going on there?

Now I have pushed fixes for MKLDNN and (hopefully) MPS. The patch for MKLDNN needs the following patch in ideep
```git
diff --git a/include/ideep/operators/prelu.hpp b/include/ideep/operators/prelu.hpp
index f958c27..6591b1f 100644
--- a/include/ideep/operators/prelu.hpp
+++ b/include/ideep/operators/prelu.hpp
@@ -15,14 +15,6 @@ struct prelu_forward : public dnnl::prelu_forward {
     auto src_in = src;
     auto weight_in = weight;

-    // Reshape weight to src dimension
-    auto new_dims = src.get_dims();
-    if (src.ndims() != weight.ndims()) {
-      std::vector<dim> dim_w(src.ndims(), 1);
-      dim_w[1] = weight.get_dim(0);
-      weight_in.reshape(dim_w);
-    }
-
     auto src_desc = src_in.get_desc();
     auto weight_desc = weight_in.get_desc().to_format_any();

@@ -61,14 +53,6 @@ struct prelu_backward : public dnnl::prelu_backward {
     auto diff_dst_in = diff_dst;
     auto weight_dims = weight_in.get_dims();

-    // Reshape wieght to src dimension
-    auto new_dims = src.get_dims();
-    if (src.ndims() != weight.ndims()) {
-      std::vector<dim> dim_w(src.ndims(), 1);
-      dim_w[1] = weight.get_dim(0);
-      weight_in.reshape(dim_w);
-    }
-
     auto src_desc = src_in.get_desc();
     auto weight_desc = weight_in.get_desc().to_format_any();
     auto diff_dst_desc = diff_dst_in.get_desc();
@@ -105,11 +89,6 @@ struct prelu_backward : public dnnl::prelu_backward {
          {DNNL_ARG_DIFF_SRC, diff_src},
          {DNNL_ARG_DIFF_WEIGHTS, diff_weight},
          {DNNL_ARG_SCRATCHPAD, scratchpad}});
-
-    // Reshape weight back to original dimension
-    if (diff_weight.get_dims() != weight_dims) {
-      diff_weight.reshape(weight_dims);
-    }
   }
 };
```
@jgong5 what's the procedure to land this into ideep?

**Comment by jbschlosser** [comment]
@lezcano Whoops, the quantization issue is that `torch.ops.quantized.prelu` still points directly to the quantized kernel (which is now registered under `_prelu_kernel`) and it no longer has the weight reshaping logic, so the broadcasting is done incorrectly. That reshaping logic is only "redundant" when entering through `torch.ops.aten.prelu`, not when the composite part is bypassed. I think we can either:

* Add back in the weight reshaping logic for the quantization kernel, accepting redundancy when invoked through the composite `prelu` OR
* Point `torch.ops.quantized.prelu` -> `prelu` instead of the kernel registered under `_prelu_kernel`. I'd defer to a quantization expert as to how to do this / if this is a bad idea

**Comment by yanbing-j** [comment]
Hi @lezcano ,

> Edit2. Note that the previous MKLDNN code was wrong in some cases if it supports 0-dim tensors. The code in core should fix that issue.

Can you explain more about the wrong cases? And why the weight reshape need to be removed, both in prelu C++ kernel and ideep? Thanks.

**Comment by lezcano** [comment]
@yanbing-j I think the code would break in the case where input is 1D and weight is 0D, then the code `dim_w[1] = weight.get_dim(0);` would be executed and would break.
This code will not be necessary with this PR, as the inputs given to all the backends would already have the same number of dimensions.

**Comment by jgong5** [comment]
> what's the procedure to land this into ideep?

No, I don't think we can do that since oneDNN expects the weight dim is the same as input. See this: https://oneapi-src.github.io/oneDNN/v2/dev_guide_prelu.html

>  I think the code would break in the case where input is 1D and weight is 0D, then the code dim_w[1] = weight.get_dim(0); would be executed and would break.

We can fix this one for the case when weight is a scalar. But it would be in the ATen/native/mkldnn/Prelu.cpp not in ideep, I guess, since ideep doesn't support 0-dim tensor. @yanbing-j

**Comment by lezcano** [comment]
If ideep does not support 0-dim tensors, then the code there was correct all along.

About the ideep fix, the thing is that ideep does some manual resizing that's no longer necessary after this PR. That code can be removed. Note that oneDNN would get the tensors with the same number of dimensions, so nothing would change there, the only thing that would change is that the resizing would now be done at the ATen level, rather than at the ideep level.

**Comment by lezcano** [comment]
Can I please get reviews from:
@kulinseth for the MPS changes
@albanD @soulitzer for the changes in the derivatives
@jbschlosser @albanD for the PR in general

I think that the quantisation and the MKLDNN changes are minimal so they don't need much of a review, or they can be reviewed by non-domain experts.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp
Comment (jbschlosser) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:684]:
```suggestion
  // hit via quantized::prelu, which is registered under quantized/cpu/qprelu.cpu
```

File: aten/src/ATen/native/native_functions.yaml
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:4657]: Do you actually need a prelu.out since this is CompositeImplicit ?

**Review by jbschlosser — APPROVED** [review]
Thanks for all your work getting this cleaned up! I'm good with it on my end though I think ideally we would avoid adding any ops.

Without adding `_prelu_kernel` / `_prelu_backward_kernel`, I think we could still get de-duplication of the broadcasting logic by dispatching CPU / CUDA to the same function and using `REGISTER_DISPATCH` dispatching from there. However, I do understand that this doesn't fix vmap / simplify writing batch rules or simplify derivative computation, which I think are valuable. So imo the tradeoffs are worth it.

File: tools/autograd/derivatives.yaml
Comment (albanD) [inline tools/autograd/derivatives.yaml:1999]: extra newline

**Review by albanD — APPROVED** [review]
SGTM

File: aten/src/ATen/native/native_functions.yaml
Comment (lezcano) [reply aten/src/ATen/native/native_functions.yaml:4657]: Leaving it as people may still want to call `torch.prelu(x, out=out)`

**Comment by lezcano** [comment]
Great coordinated reviews. Even better being the two of them approvals :D

@pytorchbot merge

**Comment by yanbing-j** [comment]
> We can fix this one for the case when weight is a scalar. But it would be in the ATen/native/mkldnn/Prelu.cpp not in ideep, I guess, since ideep doesn't support 0-dim tensor. @yanbing-j

Yes. We have changed the weight dimension to input dimension in https://github.com/pytorch/pytorch/pull/91238/files#diff-3dc8af925ec5d52a7cc5722279de60aeac5458097166318bcf6bfd58c420e629L710. The scenario of 1D input and 0D weight is supported. @jgong5 @lezcano 

> the only thing that would change is that the resizing would now be done at the ATen level, rather than at the ideep level.

Correct. The resizing has been done in ATen level, if the dimensions of src and weight are the same, the resizing in ideep level will not be invoked, and it can be regarded as a double check. @lezcano

**Comment by Xia-Weiwen** [comment]
Hi @lezcano Looks like the quantization issue is fixed, right? https://github.com/pytorch/pytorch/pull/91238#issuecomment-1366882486

**Comment by lezcano** [comment]
yeah, all good on the quantisation end. Thank you for checking in though @Xia-Weiwen

**Pull request closed without merging**
